tests/unit/utils/SGObjectIterator.h

```diff
@@ -0,0 +1,164 @@
+#ifndef __SG_OBJECT_ITERATOR_H__
+#define __SG_OBJECT_ITERATOR_H__
+
+#include <iterator>
+#include <set>
+#include <string>
+#include <shogun/base/class_list.h>
+
+using namespace shogun;
+
+// to have a type for non-template SGObject classes
+struct untemplated_sgobject
+{
+};
+
+/** Returns primitive type value for template parameter */
+template <class T>
+static EPrimitiveType sg_primitive_type();
+template <class T>
+static std::string sg_primitive_type_string();
+
+// stringizing the result of expansion of a macro argument used below
+// https://gcc.gnu.org/onlinedocs/cpp/Stringizing.html
+#define STRING__(s) #s
+#define XSTRING__(s) STRING__(s)
+
+#define SG_PRIMITIVE_TYPE(T, ptype)                                            \
+	template <>                                                                \
+	EPrimitiveType sg_primitive_type<T>()                                      \
+	{                                                                          \
+		return ptype;                                                          \
+	}                                                                          \
+                                                                               \
+	template <>                                                                \
+	std::string sg_primitive_type_string<T>()                                  \
+	{                                                                          \
+		return std::string(XSTRING__(ptype));                                  \
+	}
+
+SG_PRIMITIVE_TYPE(bool, PT_BOOL);
+SG_PRIMITIVE_TYPE(char, PT_CHAR);
+SG_PRIMITIVE_TYPE(int8_t, PT_INT8);
+SG_PRIMITIVE_TYPE(uint8_t, PT_UINT8);
+SG_PRIMITIVE_TYPE(int16_t, PT_INT16);
+SG_PRIMITIVE_TYPE(int32_t, PT_INT32);
+SG_PRIMITIVE_TYPE(int64_t, PT_INT64);
+SG_PRIMITIVE_TYPE(float32_t, PT_FLOAT32);
+SG_PRIMITIVE_TYPE(float64_t, PT_FLOAT64);
+SG_PRIMITIVE_TYPE(floatmax_t, PT_FLOATMAX);
+SG_PRIMITIVE_TYPE(untemplated_sgobject, PT_NOT_GENERIC);
+#undef SG_PRIMITIVE_TYPE
+#undef XSTRING__
+#undef STRING__
+
+/** Helper to write c++11 style loops in tests that cover all Shogun classes */
+template <typename T>
+class sg_object_iterator
+{
+public:
+	sg_object_iterator(
+	    std::set<std::string> class_names = std::set<std::string>())
+	{
+		if (class_names.empty())
```

Comment:
Id prefer offering two constructors here, one for each case, but that is a minor detail